src/mongo_odm/nvp.hpp

```diff
@@ -74,17 +58,646 @@ struct Nvp {
+    constexpr AddToSetUpdateExpr<NvpT, Iterable> add_to_set(const Iterable& iter) const {
+        return {*static_cast<const NvpT*>(this), iter, true};
+    }
+
+    /**
+     * Creates an update epxression with the $push operator, that adds a single value to an array.
+     * This is only enabled if the current field is an array.
+     * @param val   The value to add.
+     */
+    template <typename U = no_opt_type,
+              typename = typename std::enable_if<is_iterable_not_string_t<U>::value>::type>
+    constexpr PushUpdateExpr<NvpT, iterable_value_t<no_opt_type>> push(
+        const iterable_value_t<no_opt_type>& val) const {
+        return {*static_cast<const NvpT*>(this), val, false};
+    }
+
+    /**
+     * Creates an update epxression with the $push operator and the $each modifier,
+     * that adds a list of value to an array. Further modifiers can be given as arguments to this
+     * function, or by modifying the resulting PushUpdateExpr object.
+     * This is only enabled if the current field is an array.
+
+     * @tparam Iterable An iterable that contains the same value type as this field's array.
+     * @tparam Sort     The type of the given sort expression, if one is given.
+     * @param  iter     The iterable containing values to add.
+     * @param  slice    An optional argument containing the value of the $slice modifier.
+     * @param  sort    An optional argument containing an expression for the $sort modifier.
+     * @param  position    An optional argument containing the value of the $position modifier.
+     */
+    template <typename U = no_opt_type,
+              typename = typename std::enable_if<is_iterable_not_string_t<U>::value>::type,
+              typename Iterable, typename = enable_if_matching_iterable_t<Iterable>,
+              typename Sort = int,
+              typename = typename std::enable_if<details::is_sort_expression<Sort>::value ||
+                                                 std::is_same<int, Sort>::value>::type>
+    constexpr PushUpdateExpr<NvpT, Iterable, Sort> push(
+        const Iterable& iter, bsoncxx::stdx::optional<std::int32_t> slice = bsoncxx::stdx::nullopt,
+        const bsoncxx::stdx::optional<Sort>& sort = bsoncxx::stdx::nullopt,
+        bsoncxx::stdx::optional<std::uint32_t> position = bsoncxx::stdx::nullopt) const {
+        return {*static_cast<const NvpT*>(this), iter, true, slice, sort, position};
+    }
 };
 
+/**
+ * A type trait struct that inherits from std::true_type
+ * if the given type parameter is a name-value pair,
+ * and from std::false_type otherwise.
+ */
+template <typename>
+struct is_nvp_type : public std::false_type {};
+
+template <typename Base, typename T>
+struct is_nvp_type<Nvp<Base, T>> : public std::true_type {};
+
+template <typename Base, typename T, typename Parent>
+struct is_nvp_type<NvpChild<Base, T, Parent>> : public std::true_type {};
+
+template <typename T>
+struct is_nvp_type<FreeNvp<T>> : public std::true_type {};
+
```

Comment:
maybe add a `using is_nvp_v = is_nvp<T>`. I think the`_v` idiom is getting more common